libcontainer/nsenter/nsexec.c

```diff
@@ -191,22 +194,83 @@ static void update_setgroups(int pid, enum policy_t setgroup)
 	}
 }
 
-static void update_uidmap(int pid, char *map, size_t map_len)
+static int try_mapping_tool(const char *app, int pid, char *map, size_t map_len)
+{
+	int child = fork();
+	if (child < 0)
+		bail("failed to fork");
+
+	if (child == 0) {
+#define MAX_ARGV 20
+		char *argv[MAX_ARGV];
+		char pid_fmt[16];
+		int argc = 0;
+		size_t i;
+
+		sprintf (pid_fmt, "%d", pid);
+
+		argv[argc++] = (char *) app;
+		argv[argc++] = pid_fmt;
+		/*
+		 * Convert the map string into a list of argument that
+		 * newuidmap/newgidmap can understand.
+		 */
+		for (i = 0; i < map_len - 1; ) {
+			argv[argc++] = &map[i++];
+
+			while (i < map_len - 1 && map[i] != '\n'
+	                       && map[i] != ' ' && map[i] != '\0')
+				i++;
+			map[i++] = '\0';
+
+			while (i < map_len - 1 && map[i] == '\n' && map[i] == ' ')
+				i++;
+			if (map[i] == '\0')
+				break;
+		}
+		execvp (app, argv);
+	}
+	else {
+		int status;
+		while (true) {
+			if (waitpid(child, &status, 0) < 0) {
+				if (errno == EINTR)
+					continue;
+				bail("failed to waitpid");
+			}
+			if (WIFEXITED(status))
+				return WEXITSTATUS(status);
+                  }
+	}
+	return -1;
+}
+
+static void update_uidmap(int pid, uint8_t is_rootless, char *map, size_t map_len)
 {
 	if (map == NULL || map_len <= 0)
 		return;
 
-	if (write_file(map, map_len, "/proc/%d/uid_map", pid) < 0)
+	if (write_file(map, map_len, "/proc/%d/uid_map", pid) < 0) {
+		if(errno == EPERM
+                   && try_mapping_tool ("newuidmap", pid, map, map_len) == 0) {
+			return;
+		}
 		bail("failed to update /proc/%d/uid_map", pid);
+	}
 }
 
-static void update_gidmap(int pid, char *map, size_t map_len)
+static void update_gidmap(int pid, uint8_t is_rootless, char *map, size_t map_len)
 {
 	if (map == NULL || map_len <= 0)
 		return;
 
-	if (write_file(map, map_len, "/proc/%d/gid_map", pid) < 0)
+	if (write_file(map, map_len, "/proc/%d/gid_map", pid) < 0) {
+		if(errno == EPERM
+                   && try_mapping_tool ("newgidmap", pid, map, map_len) == 0) {
+			return;
+		}
 		bail("failed to update /proc/%d/gid_map", pid);
+	}
```

Comment:
Completely stylistic, but I think this is more readable (same applies for `update_uidmap`):

```
if (write_file(map, map_len, "/proc/%d/gid_map", pid) < 0) {
	if (errno != EPERM)
		bail("failed to update /proc/%d/gid_map", pid);
	if (try_mapping_tool("newgidmap", pid, map, map_len))
		bail("failed to use newgid map on %d", pid);
}
```